autocorns/biologist.py

```diff
@@ -98,149 +127,137 @@ def unicorn_metadata(
+                )
+                tokens_metadata.extend(make_multicall_result)
+                calls_progress_bar.update(len(tokens_ids_chunk))
+                break
+            except ValueError:
+                time.sleep(1)
                 continue
-            submission_progress_bar.update()
-            future = executor.submit(
-                contract.get_unicorn_metadata, token_id, block_number
-            )
-            jobs[future] = token_id
 
-    for future in as_completed(jobs, timeout):
-        metadata_progress_bar.update()
-        token_id = jobs[future]
+    for token_id, token_data in zip(token_ids, tokens_metadata):
         try:
-            metadata = future.result()
             result = {
                 "token_id": token_id,
                 "block_number": block_number,
-                "lifecycle_stage": metadata[3],
-                "class_number": metadata[-2],
+                "lifecycle_stage": token_data[3],
+                "class_number": token_data[-2],
             }
             results.append(result)
         except Exception as e:
             error = {
                 "token_id": token_id,
                 "block_number": block_number,
-                "error": f"Failed to retrieve DNA: {str(e)}",
+                "error": f"Failed retrive unicorns metadata: {str(e)}",
             }
             errors.append(error)
-
     return results, errors
 
 
 def unicorn_mythic_body_parts(
     contract_address: ChecksumAddress,
     dnas: List[Dict[str, Any]],
     block_number: Optional[int] = None,
-    num_workers: int = 1,
-    timeout: float = 30.0,
-    checkpoint_file: Optional[str] = None,
 ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
     if block_number is None:
         block_number = len(chain) - 1
 
     results: List[Dict[str, Any]] = []
     errors: List[Dict[str, Any]] = []
 
-    dnas_index = {item["token_id"]: item for item in dnas}
-
-    checkpointed_token_ids: Set[int] = set()
-    if checkpoint_file is not None:
-        with open(checkpoint_file, "r") as ifp:
-            for line in ifp:
-                stripped_line = line.strip()
-                if stripped_line:
-                    result = json.loads(stripped_line)
-                    if result["token_id"] not in dnas_index:
-                        continue
-                    dna_result = dnas_index[result["token_id"]]
-                    if (
-                        result.get("num_mythic_body_parts") is not None
-                        and result.get("dna") != "2"
-                        and dna_result["block_number"] <= result["block_number"]
-                    ):
-                        results.append(result)
-                        checkpointed_token_ids.add(result["token_id"])
-
-    submission_progress_bar = tqdm(
-        total=len(dnas) - len(checkpointed_token_ids),
-        desc="Submitting requests for number of unicorn mythic body parts",
-    )
+    tokens_metadata = []
```

Comment:
(Related to comment below) You can make this a dictionary instead of an array. Map token IDs to their metadata.

Reply:
Do not simplify work in next steps of nerge can skip.